readthedocs/oauth/services/bitbucket.py

```diff
@@ -0,0 +1,210 @@
+        # existing repositories.
+        try:
+            resp = self.paginate(
+                'https://bitbucket.org/api/2.0/repositories/?role=admin')
+            repos = (
+                RemoteRepository.objects
+                .filter(users=self.user,
+                        full_name__in=[r['full_name'] for r in resp],
+                        account=self.account)
+            )
+            for repo in repos:
+                repo.admin = True
+                repo.save()
+        except (TypeError, ValueError):
+            pass
+
+    def sync_teams(self):
+        """Sync Bitbucket teams and team repositories"""
+        try:
+            teams = self.paginate(
+                'https://api.bitbucket.org/2.0/teams/?role=member'
+            )
+            for team in teams:
+                org = self.create_organization(team)
+                repos = self.paginate(team['links']['repositories']['href'])
+                for repo in repos:
+                    self.create_repository(repo, organization=org)
+        except ValueError as e:
+            log.error('Error syncing Bitbucket organizations: %s',
+                      str(e), exc_info=True)
+            raise Exception('Could not sync your Bitbucket team repositories, '
+                            'try reconnecting your account')
+
+    def create_repository(self, fields, privacy=DEFAULT_PRIVACY_LEVEL,
+                          organization=None):
+        """Update or create a repository from Bitbucket API response
+
+        .. note::
+            The :py:data:`admin` property is not set during creation, as
+            permissions are not part of the returned repository data from
+            Bitbucket.
+
+        :param fields: dictionary of response data from API
+        :param privacy: privacy level to support
+        :param organization: remote organization to associate with
+        :type organization: RemoteOrganization
+        :rtype: RemoteRepository
+        """
+        if (fields['is_private'] is True and privacy == 'private' or
+                fields['is_private'] is False and privacy == 'public'):
+            repo, _ = RemoteRepository.objects.get_or_create(
+                full_name=fields['full_name'],
+                account=self.account,
+            )
+            if repo.organization and repo.organization != organization:
+                log.debug('Not importing %s because mismatched orgs' %
+                          fields['name'])
+                return None
+            else:
+                repo.organization = organization
+            repo.users.add(self.user)
+            repo.name = fields['name']
+            repo.description = fields['description']
+            repo.private = fields['is_private']
+            for link in fields['links']['clone']:
+                if link['name'] == 'https':
+                    repo.clone_url = link['href']
+                    # Remove user name from HTTPS clone url
```

Comment:
Oh, this is to ensure a url in the form: `http://bitbucket.org/user/project` instead of: `http://user@bitbucket.org/user/project`, as linked to in the API response. We do a check on the import form for `@` to make a guess as to whether the repo is private or not. I meant to open a ticket that will address linking the `RemoteRepository` to the `Project` object, and using the API response during import instead.